src/core_plugins/kibana/public/dashboard/panel/panel_header/panel_actions/build_context_menu.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * Loops through allActions and extracts those that belong on the given contextMenuPanelId
+ * @param {string} contextMenuPanelId
+ * @param {Array.<DashboardPanelAction>} allActions
+ */
+function getActionsForPanel(contextMenuPanelId, allActions) {
+  const panelActions = [];
+  for (let i = 0; i < allActions.length; i++) {
+    const action = allActions[i];
+    if (action.parentPanelId === contextMenuPanelId) {
+      panelActions.push(action);
+    }
+  }
+  return panelActions;
+}
+
+/**
+ * @param {String} contextMenuPanelId
+ * @param {Array.<DashboardPanelAction>} actions
+ * @param {Embeddable} embeddable
+ * @param {ContainerState} containerState
+ * @return {{
+ *   Array.<EuiContextMenuPanelItemShape> items - panel actions converted into the items expected to be on an
+ *     EuiContextMenuPanel,
+ *   Array.<EuiContextMenuPanelShape> childPanels - extracted child panels, if any actions also open a panel. They
+ *     need to be moved to the top level for EUI.
+ *  }}
+ */
+function buildEuiContextMenuPanelItemsAndChildPanels({ contextMenuPanelId, actions, embeddable, containerState }) {
+  const items = [];
+  let childPanels = [];
+  const actionsForPanel = getActionsForPanel(contextMenuPanelId, actions);
+  actionsForPanel.forEach(action => {
+    const isVisible = action.isVisible({ embeddable, containerState });
+    if (!isVisible) {
+      return;
+    }
+
+    let childPanelToOpenOnClick;
+    if (action.childContextMenuPanel) {
+      childPanels = childPanels.concat(
```

Comment:
Why are you assigning childPanels instead of just using push and array deconstructing?